### PR TITLE
Testsuite code refactor

### PR DESCRIPTION
**Pull request by Jiawei0227**
Extract TestSuite, TestDriver, TestPattern, TestConfig
and VolumeResource, SnapshotVolumeResource from testsuite
package and put them into a new package called api.

The ultimate goal here is to make the testsuites as clean
as possible. And only testsuites in the package.

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Right now there are too many different things in the testsuite package which makes it really hard to read and understand.

This PR is to code refactor the `test/e2e/storage/testsuite` package to extract all the public type definition out and move to a new package called `test/e2e/storage/api`. The ultimate goal here is to keep the testsuite package as clean and easy to extend as possible. And also isolate the type definition so that we can make the testsuite pkg no different than a third party package. This also makes sure a third party user is easy to use these code and write their own testsuite.

There are also some utility move around in this PR which is relevant to keep the pkg dependency clean and right.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:
This is a first step to the testsuite cleanup. We will need to reduce redundant code after this and make sure the utility we write is not building the same wheel again and again.

**Does this PR introduce a user-facing change?**:

```release-note
storage related e2e testsuite redesign
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs
NONE
```

/sig storage
/sig testing

/cc @msau42 
/cc @jingxu97 
/cc @chrishenzie 
/cc @xing-yang 
/cc @saad-ali

### CODE REVIEW

File: test/e2e/storage/api/testsuite.go
Comment (jingxu97) [inline test/e2e/storage/api/testsuite.go:56]: what is the difference between these two skips?

File: test/e2e/storage/api/testsuite.go
Comment (jingxu97) [inline test/e2e/storage/api/testsuite.go:85]: there is no TestSuiteHandler, right?

File: test/e2e/storage/api/testsuite.go
Comment (jingxu97) [inline test/e2e/storage/api/testsuite.go:121]: it's not your change. but for some test patterns, it does not specify a fsType which means it is default fsType. does it is considered as supported or not?

File: test/e2e/storage/api/testsuite.go
Comment (jingxu97) [inline test/e2e/storage/api/testsuite.go:128]: I am thinking after I separate in-tree windows gce pd driver, we don't need this check. A driver should specify what fsType it support.

Comment (msau42) [reply test/e2e/storage/api/testsuite.go:128]: We need to check other intree cloud drivers too. It's also possible aws and azure are running these e2es outside of Kubernetes. cc @andyzhangx @wongma7 @jsafrane

**Comment by Jiawei0227** [comment]
I was wrong about using private method in the TestSuite interface. It turns out that is not possible. If anyone has any good suggestion please let me know.
https://stackoverflow.com/questions/26181271/is-it-possible-to-implement-an-interface-with-unexported-methods-in-another-pack
This has an explanation.

File: test/e2e/storage/api/testsuite.go
Comment (Jiawei0227) [reply test/e2e/storage/api/testsuite.go]: The first is globally we filter out some invalid combination. So something like if the pattern ask for inlineVolume but the driver does not support inline at all. It will skip.

The second skip is a testsuite specific skip. Each testsuite will have some certain test patterns or drivers that it does not support to run. So we can add those there.

File: test/e2e/storage/api/testsuite.go
Comment (Jiawei0227) [reply test/e2e/storage/api/testsuite.go]: Yep, that is removed per @chrishenzie suggest.

File: test/e2e/storage/api/testsuite.go
Comment (Jiawei0227) [reply test/e2e/storage/api/testsuite.go:120]: Hmm good question. It seems when we initialize Driver we already add ""
https://github.com/kubernetes/kubernetes/blob/da75c266487c52425d3eb07df30229aea04bb28e/test/e2e/storage/drivers/csi.go#L86
The other drivers in this file also correct. So I think as long as we correctly init the driver it should all support ""

File: test/e2e/storage/external/testdata/example.yaml
Comment (chrishenzie) [inline test/e2e/storage/external/testdata/example.yaml:16]: Can you include `snapshotDataSource: true`?

File: test/e2e/storage/external/testdata/example.yaml
Comment (chrishenzie) [inline test/e2e/storage/external/testdata/example.yaml:19]: Can you include `NumSnapshots: 10`? (Or some reasonable value), this will help users to run the volume snapshot stress test.

File: test/e2e/framework/volume/fixtures.go
Comment (chrishenzie) [inline test/e2e/framework/volume/fixtures.go:719]: Why are these pod exec functions a part of this volume package? From what I can tell, it does not contain any volume-specific logic. Is test/e2e/storage/utils/pod.go a better location? (I'm not sure, just spitballing an idea).

**Review by chrishenzie — CHANGES_REQUESTED** [review]
Thank you for this cleanup, this will make handling weird driver / test pattern combination skipping much simpler.

Do we have any clear docs (or docstrings) regarding what belongs in the `storageapi` package, and what doesn't? I have a similar question for the `storageutil` package. An example of this is `CreateVolumeSource()`, this currently lives in `storageutil`. Should it stay there, or should it be part of `storageapi`? I am not sure.

Maybe I'm being too cautious, but I just want to be careful that neither of these packages become easy locations to place any miscellaneous functions that are not a test suite when we could organize them further into sub-packages.

File: test/e2e/storage/api/testsuite.go
Comment (jingxu97) [reply test/e2e/storage/api/testsuite.go:127]: I think using fstype and OS to check is not ideal considering in the future, same fstype might be supported by both linux and windows for some drivers.

Not part of your change. I am proposing to have supportedOSes in driver configuration. It explicitly specify what OSes it can support and the testsuits will check based on framework.NodeDistrol value. The default value can be set to "linux".

File: test/e2e/storage/api/testsuite.go
Comment (jingxu97) [reply test/e2e/storage/api/testsuite.go:120]: this design means that for certain feature/capability, if no fsType is set for that feature, we can only test default fsType. For drivers that do not specify default fstype (empty string), that means this test will be skipped? 

But for snapshottable tests, I see external gcepd driver does not specify default fstype, but the tests still be triggered. https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/blob/master/test/k8s-integration/driver-config.go#L71

**Comment by chrishenzie** [comment]
Thinking this through a little more, it looks like there are two independent changes going on in this PR. I see one change is cleaning up how we skip unsupported tests for our test suites, and the other change is extracting functions into better locations to keep things organized and clean.

I'm wondering if it would make sense to break this down into two separate commits (they could still be part of this PR), which would make reviewing these changes easier. I am concerned that I may have glossed over something in my review that could cause a problem. Let me know what your thoughts are.

File: test/e2e/storage/api/testconfig.go
Comment (Jiawei0227) [reply test/e2e/storage/api/testconfig.go:67]: oops, I should be the end.

File: test/e2e/storage/external/testdata/example.yaml
Comment (Jiawei0227) [reply test/e2e/storage/external/testdata/example.yaml:16]: Done

File: test/e2e/storage/external/testdata/example.yaml
Comment (Jiawei0227) [reply test/e2e/storage/external/testdata/example.yaml:19]: Done

File: test/e2e/framework/volume/fixtures.go
Comment (Jiawei0227) [reply test/e2e/framework/volume/fixtures.go:719]: There is a dependency issue here.
PodExec is used by "VerifyExecInPodSucceed" and "VerifyExecInPodFail" which are used by "CheckVolumeModeOfPath".

And this "CheckVolumeModeOfPath" is also used by e2evolume. So ideally we dont want to put them under storage because that will have a situation where parent package import subpackage.

I think ideally these function should rest in e2epod. But then there will be another dependency issue happen which is caused by : https://github.com/kubernetes/kubernetes/issues/81245

I think I will put this PodExec here for now and had another separate PR maybe to clear the pods dependency in framework so that we can move this over.

**Comment by Jiawei0227** [comment]
> I'm wondering if it would make sense to break this down into two separate commits (they could still be part of this PR), which would make reviewing these changes easier. I am concerned that I may have glossed over something in my review that could cause a problem. Let me know what your thoughts are. If this is not possible for this PR then that is ok.

Yep, that totally makes sense... But given the situation now I am in... It becomes a little hard to separate the commit now.... I will make sure the following commits that I had separate things up and I can squash them after I got approval.

**Comment by Jiawei0227** [comment]
> I'm wondering if it would make sense to break this down into two separate commits (they could still be part of this PR), which would make reviewing these changes easier. I am concerned that I may have glossed over something in my review that could cause a problem. Let me know what your thoughts are. If this is not possible for this PR then that is ok.

Actually... I think you are right. I will give it a shot and separate the commits.

Otherwise it is just impossible to review. I will create a new PR.

**Comment by Jiawei0227** [comment]
The new PR is here: https://github.com/kubernetes/kubernetes/pull/96573

**Pull request closed without merging**
